File: packages/shared/src/services/alerts.ts
import { getApiClient } from './apiClient';
import { PriceAlert, AlertCreateRequest, AlertUpdateRequest } from '../types/alerts';

function convertApiAlert(apiAlert: any): PriceAlert {
  return {
    id: apiAlert.id,
    userId: apiAlert.user_id,
    dealerName: apiAlert.dealer_name,
    scriptName: apiAlert.script_name,
    condition: apiAlert.condition,
    rateType: apiAlert.rate_type,
    threshold: apiAlert.threshold,
    isActive: apiAlert.is_active,
    triggerMode: apiAlert.trigger_mode,
    cooldownMinutes: apiAlert.cooldown_minutes,
    lastTriggeredAt: apiAlert.last_triggered_at,
    createdAt: apiAlert.created_at,
    updatedAt: apiAlert.updated_at,
  };
}

export const alertService = {
  async getAlerts(): Promise<PriceAlert[]> {
    try {
      const api = getApiClient();
      const response = await api.get('/api/alerts');
      return response.data.map(convertApiAlert);
    } catch (error: any) {
      console.error('Failed to fetch alerts:', error);
      if (error.response?.data?.detail) {
        throw new Error(error.response.data.detail);
      }
      throw new Error('Failed to fetch alerts');
    }
  },

  async createAlert(data: AlertCreateRequest): Promise<PriceAlert> {
    try {
      const api = getApiClient();
      const response = await api.post('/api/alerts', {
        dealer_name: data.dealerName,
        script_name: data.scriptName,
        condition: data.condition,
        rate_type: data.rateType,
        threshold: data.threshold,
        trigger_mode: data.triggerMode || 'one_shot',
        cooldown_minutes: data.cooldownMinutes || 30,
      });
      return convertApiAlert(response.data);
    } catch (error: any) {
      console.error('Failed to create alert:', error);
      if (error.response?.data?.detail) {
        throw new Error(error.response.data.detail);
      }
      throw new Error('Failed to create alert');
    }
  },

  async updateAlert(alertId: string, data: AlertUpdateRequest): Promise<PriceAlert> {
    try {
      const api = getApiClient();
      const payload: any = {};
      if (data.threshold !== undefined) payload.threshold = data.threshold;
      if (data.condition !== undefined) payload.condition = data.condition;
      if (data.rateType !== undefined) payload.rate_type = data.rateType;
      if (data.isActive !== undefined) payload.is_active = data.isActive;
      if (data.triggerMode !== undefined) payload.trigger_mode = data.triggerMode;
      if (data.cooldownMinutes !== undefined) payload.cooldown_minutes = data.cooldownMinutes;

      const response = await api.put(`/api/alerts/${alertId}`, payload);
      return convertApiAlert(response.data);
    } catch (error: any) {
      console.error('Failed to update alert:', error);
      if (error.response?.data?.detail) {
        throw new Error(error.response.data.detail);
      }
      throw new Error('Failed to update alert');
    }
  },

  async deleteAlert(alertId: string): Promise<void> {
    try {
      const api = getApiClient();
      await api.delete(`/api/alerts/${alertId}`);
    } catch (error: any) {
      console.error('Failed to delete alert:', error);
      if (error.response?.data?.detail) {
        throw new Error(error.response.data.detail);
      }
      throw new Error('Failed to delete alert');
    }
  },
};

export default alertService;

File: backend/app/api/alerts.py
"""
Price Alert API endpoints
Provides CRUD operations for user price alerts
"""

from fastapi import APIRouter, HTTPException, Depends, status
from typing import List
import logging

from app.database.models import User
from app.services.alert_service import alert_service
from app.schemas.alerts import AlertCreate, AlertUpdate, AlertResponse
from .auth import require_subscription

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/alerts", tags=["alerts"])


@router.get("", response_model=List[AlertResponse])
async def list_alerts(current_user: User = Depends(require_subscription)):
    try:
        return await alert_service.get_alerts(current_user.id)
    except Exception as e:
        logger.error(f"Error fetching alerts for user {current_user.id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch alerts"
        )


@router.post("", response_model=AlertResponse, status_code=status.HTTP_201_CREATED)
async def create_alert(
    data: AlertCreate,
    current_user: User = Depends(require_subscription)
):
    try:
        return await alert_service.create_alert(current_user.id, data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    except Exception as e:
        logger.error(f"Error creating alert for user {current_user.id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create alert"
        )


@router.put("/{alert_id}", response_model=AlertResponse)
async def update_alert(
    alert_id: str,
    data: AlertUpdate,
    current_user: User = Depends(require_subscription)
):
    try:
        return await alert_service.update_alert(current_user.id, alert_id, data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except Exception as e:
        logger.error(f"Error updating alert {alert_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update alert"
        )


@router.delete("/{alert_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_alert(
    alert_id: str,
    current_user: User = Depends(require_subscription)
):
    try:
        await alert_service.delete_alert(current_user.id, alert_id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(e))
    except Exception as e:
        logger.error(f"Error deleting alert {alert_id}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete alert"
        )
